File: src/vibeusage/cli/commands/cache.py
"""Cache management commands for vibeusage."""

from pathlib import Path

import typer
from rich.console import Console
from rich.table import Table

from vibeusage.cli.app import app, ExitCode
from vibeusage.cli.atyper import ATyper
from vibeusage.config.cache import (
    cache_org_id,
    cache_snapshot,
    clear_all_cache,
    clear_org_id_cache,
    clear_provider_cache,
    is_snapshot_fresh,
    load_cached_snapshot,
    org_id_path,
    snapshot_path,
)
from vibeusage.config.paths import cache_dir, snapshots_dir
from vibeusage.config.settings import get_config
from vibeusage.providers import list_provider_ids

# Create cache group
cache_app = ATyper(help="Manage cached usage data.")


@cache_app.callback(invoke_without_command=True)
def cache_callback(
    ctx: typer.Context,
) -> None:
    """Cache management commands.

    If no subcommand is provided, shows cache status.
    """
    # Store context for subcommands
    pass


@cache_app.command("show")
def cache_show_command(
    ctx: typer.Context,
) -> None:
    """Show cache status per provider."""
    console = Console()

    config = get_config()
    stale_threshold = config.fetch.stale_threshold_minutes
    json_mode = ctx.meta.get("json", False)
    verbose = ctx.meta.get("verbose", False)
    quiet = ctx.meta.get("quiet", False)

    # Build cache status data
    cache_data = {}
    for provider_id in list_provider_ids():
        snap_path = snapshot_path(provider_id)
        org_path = org_id_path(provider_id)

        snap_status = "none"
        snap_age_minutes = None
        org_status = False

        if snap_path.exists():
            snapshot = load_cached_snapshot(provider_id)
            if snapshot:
                if is_snapshot_fresh(snapshot, stale_threshold):
                    snap_status = "fresh"
                else:
                    snap_status = "stale"

                from datetime import datetime, timezone

                now = datetime.now(timezone.utc)
                if snapshot.fetched_at.tzinfo is None:
                    fetched = snapshot.fetched_at.replace(tzinfo=timezone.utc)
                else:
                    fetched = snapshot.fetched_at

                delta = now - fetched
                snap_age_minutes = int(delta.total_seconds() / 60)
            else:
                snap_status = "error"

        if org_path.exists():
            org_status = True

        cache_data[provider_id] = {
            "snapshot": snap_status,
            "org_id_cached": org_status,
            "age_minutes": snap_age_minutes,
        }

    if json_mode:
        from vibeusage.display.json import output_json_pretty

        output_json_pretty(cache_data)
        return

    # Quiet mode: minimal output
    if quiet:
        for provider_id, data in cache_data.items():
            console.print(f"{provider_id}: {data['snapshot']}")
        return

    # Display table
    table = Table(title="Cache Status", show_header=True, header_style="bold")
    table.add_column("Provider", style="cyan")
    table.add_column("Snapshot", style="yellow")
    table.add_column("Org ID", style="dim")
    table.add_column("Age", style="dim")

    for provider_id, data in cache_data.items():
        snap_status_map = {
            "fresh": "[green]✓ Fresh[/green]",
            "stale": "[yellow]⚠ Stale[/yellow]",
            "error": "[red]✗ Error[/red]",
            "none": "—",
        }
        snap_status = snap_status_map.get(data["snapshot"], "—")

        org_status = "[green]✓[/green]" if data["org_id_cached"] else "—"

        age = data["age_minutes"]
        if age is None:
            snap_age = "—"
        elif age >= 1440:
            snap_age = f"{age // 1440}d"
        elif age >= 60:
            snap_age = f"{age // 60}h"
        elif age >= 1:
            snap_age = f"{age}m"
        else:
            snap_age = "<1m"

        table.add_row(provider_id, snap_status, org_status, snap_age)

    console.print(table)
    console.print(f"\nCache directory: {cache_dir()}")

    # Verbose: show stale threshold
    if verbose:
        console.print(f"\n[dim]Stale threshold: {stale_threshold} minutes[/dim]")
        console.print(f"[dim]Snapshots older than this are considered stale.[/dim]")


@cache_app.command("clear")
def cache_clear_command(
    ctx: typer.Context,
    provider: str = typer.Argument(
        None,
        help="Provider to clear cache for (default: all)",
    ),
    org_only: bool = typer.Option(
        False,
        "--org",
        "-o",
        help="Only clear org ID cache",
    ),
) -> None:
    """Clear cache data."""
    console = Console()
    json_mode = ctx.meta.get("json", False)

    result = {"success": True, "cleared": []}

    if provider is None:
        # Clear all cache
        if org_only:
            clear_all_cache(org_ids_only=True)
            result["message"] = "Cleared all org ID cache"
        else:
            clear_all_cache()
            result["message"] = "Cleared all cache"
        result["provider"] = "all"
        result["org_only"] = org_only
    else:
        if provider not in list_provider_ids():
            result["success"] = False
            result["error"] = f"Unknown provider: {provider}"
            if json_mode:
                from vibeusage.display.json import output_json_pretty
                output_json_pretty(result)
            console.print(f"[red]Unknown provider:[/red] {provider}")
            console.print(f"Available providers: {', '.join(list_provider_ids())}")
            raise typer.Exit(ExitCode.CONFIG_ERROR)

        result["provider"] = provider
        result["org_only"] = org_only
        result["cleared"].append(provider)

        if org_only:
            clear_org_id_cache(provider)
            result["message"] = f"Cleared org ID cache for {provider}"
        else:
            clear_provider_cache(provider)
            result["message"] = f"Cleared cache for {provider}"

    if json_mode:
        from vibeusage.display.json import output_json_pretty
        output_json_pretty(result)
        return

    if result["success"]:
        console.print(f"[green]✓[/green] {result['message']}")


# Register the cache group with the main app
app.add_typer(cache_app, name="cache")

File: vibeusage-go/cmd/cache.go
package cmd

import (
	"fmt"
	"os"
	"sort"
	"strings"
	"time"

	"github.com/spf13/cobra"

	"github.com/joshuadavidthomas/vibeusage/internal/config"
	"github.com/joshuadavidthomas/vibeusage/internal/display"
	"github.com/joshuadavidthomas/vibeusage/internal/provider"
)

var cacheCmd = &cobra.Command{
	Use:   "cache",
	Short: "Manage cached usage data",
}

var cacheShowCmd = &cobra.Command{
	Use:   "show",
	Short: "Show cache status per provider",
	RunE: func(cmd *cobra.Command, args []string) error {
		cfg := config.Get()
		staleThreshold := cfg.Fetch.StaleThresholdMinutes

		ids := provider.ListIDs()
		sort.Strings(ids)

		type cacheInfo struct {
			Snapshot string `json:"snapshot"`
			OrgID    bool   `json:"org_id_cached"`
			Age      *int   `json:"age_minutes"`
		}

		cacheData := make(map[string]cacheInfo)

		for _, pid := range ids {
			info := cacheInfo{Snapshot: "none"}

			snap := config.LoadCachedSnapshot(pid)
			if snap != nil {
				age := int(time.Since(snap.FetchedAt).Minutes())
				info.Age = &age
				if age < staleThreshold {
					info.Snapshot = "fresh"
				} else {
					info.Snapshot = "stale"
				}
			}

			orgPath := config.OrgIDPath(pid)
			if _, err := os.Stat(orgPath); err == nil {
				info.OrgID = true
			}

			cacheData[pid] = info
		}

		if jsonOutput {
			display.OutputJSON(cacheData)
			return nil
		}

		if quiet {
			for _, pid := range ids {
				fmt.Printf("%s: %s\n", pid, cacheData[pid].Snapshot)
			}
			return nil
		}

		fmt.Println("Cache Status")
		fmt.Println(strings.Repeat("─", 50))
		fmt.Printf("%-12s %-12s %-8s %s\n", "Provider", "Snapshot", "Org ID", "Age")
		fmt.Println(strings.Repeat("─", 50))

		for _, pid := range ids {
			info := cacheData[pid]
			snapStatus := "—"
			switch info.Snapshot {
			case "fresh":
				snapStatus = "✓ Fresh"
			case "stale":
				snapStatus = "⚠ Stale"
			}

			orgStatus := "—"
			if info.OrgID {
				orgStatus = "✓"
			}

			ageStr := "—"
			if info.Age != nil {
				a := *info.Age
				if a >= 1440 {
					ageStr = fmt.Sprintf("%dd", a/1440)
				} else if a >= 60 {
					ageStr = fmt.Sprintf("%dh", a/60)
				} else if a >= 1 {
					ageStr = fmt.Sprintf("%dm", a)
				} else {
					ageStr = "<1m"
				}
			}

			fmt.Printf("%-12s %-12s %-8s %s\n", pid, snapStatus, orgStatus, ageStr)
		}

		fmt.Printf("\nCache directory: %s\n", config.CacheDir())
		return nil
	},
}

var cacheClearCmd = &cobra.Command{
	Use:   "clear [provider]",
	Short: "Clear cache data",
	RunE: func(cmd *cobra.Command, args []string) error {
		var providerID string
		if len(args) > 0 {
			providerID = args[0]
			if _, ok := provider.Get(providerID); !ok {
				return fmt.Errorf("unknown provider: %s", providerID)
			}
		}

		config.ClearAllCache(providerID)

		msg := "Cleared all cache"
		if providerID != "" {
			msg = fmt.Sprintf("Cleared cache for %s", providerID)
		}

		if jsonOutput {
			display.OutputJSON(map[string]any{
				"success":  true,
				"message":  msg,
				"provider": providerID,
			})
			return nil
		}

		fmt.Printf("✓ %s\n", msg)
		return nil
	},
}

func init() {
	cacheCmd.AddCommand(cacheShowCmd)
	cacheCmd.AddCommand(cacheClearCmd)
}
